### DISCUSSION TITLE
How to reset previous focus when moving away from Menu

### DISCUSSION THREAD

**Discussion by bengry** [ariakit/ariakit#3635] [Questions] [answered]
I'm using a `<Menubar>` and `<MenuItem>` to implement a side-navigation, and seeing some behavior I can't find a way to disable - when <kbd>Tab</kbd>-ing away from the menubar and going back, the focus is kept on the last focused item.
Is there a way to disable this behavior?

This is best explained by this short video:
<video src="https://github.com/ariakit/ariakit/assets/260431/b0f190c1-6b01-4424-814c-bdc3619db7a3">

[Stackblitz reproduction](https://stackblitz.com/edit/vitejs-vite-qtvagp?file=src%2FApp.tsx)

I'd like to have the focus on "About" when I tab back from `header focusable element`, not on `Blog`, which was the last focused item.

As a side question, is `<Menubar>` (and `<MenuItem>`) the best way to approach this layout the way I have it in the Stackblitz?

Lastly, what's the purpose of the `<MenuProvider>` here? Is it required?

**Comment by diegohaz** [comment]
I don't think you should use `Menubar` for that type of navigation. See [Can I use Menubar for site navigation?](https://ariakit.org/components/menubar#can-i-use-menubar-for-site-navigation).

Simply use good old semantic HTML:

```jsx
<nav>
  <ul>
    <li><a href="/about">About</a></li>
    <li><a href="/contact">Contact</a></li>
    <li><a href="/blog">Blog</a></li>
    ...
  </ul>
</nav>
```

**Reply by bengry** [reply]
But then the user has the tab all the way through the menu items just to reach the header, instead of giving them the option to navigate within the menubar using the arrow keys, yet tab straight away to the header using <kbd>Tab</kbd>. Is that not better?

**Reply by diegohaz** [reply]
It's better to keep navigation links tabbable. You can use [Skip Navigation Links](https://webaim.org/techniques/skipnav/) to allow keyboard users to skip to the main content.

**Reply by bengry** [reply]
Thanks, I appreciate the help with your knowledge of aria concerns. I'll change it to a regular `<ul>`/`<li>`.

Just for my own curiosity's sake - is the behavior I asked about disable-able? I noticed it also exists in Tabs (and probably other components in Ariakit).

**Reply by diegohaz** [reply, marked as answer]
That's typically how [composite widgets](https://ariakit.org/components/composite) like `menubar` and `tablist` work. Though you can reset the `activeId` state on blur if necessary:

```jsx
const store = useMenubarStore();

<Menubar
  store={store}
  onBlur={(event) => {
    // ignore blur events inside the menubar
    if (event.currentTarget.contains(event.relatedTarget)) return;
    // reset the active item id
    store.setActiveId(undefined);
  }}
```

**Reply by bengry** [reply]
Thanks, that works great.

What about:
> Lastly, what's the purpose of the <MenuProvider> here? Is it required?

And on a bit of an off-topic, but might be helpful to others - would [your recommendation](https://github.com/ariakit/ariakit/discussions/3635#discussioncomment-8932285) still hold true if these top-level items would open or be related to a drawer of nested items? e.g.
![CleanShot 2024-03-27 at 21 27 25@2x](https://github.com/ariakit/ariakit/assets/260431/a96d08eb-0d4a-4837-9641-dbec35473385)

i.e. the About/Contact/Store/Solutions do a navigation, but also have a related element of sub-pages. In the picture - the Store item has sub-sections you can navigate to.

(I updated the [Stackblitz](https://stackblitz.com/edit/vitejs-vite-qtvagp?file=src%2FApp.tsx), in case you want to take a look at the rough layout, although the CSS there is hacky, just to quickly get an example up).

**Reply by diegohaz** [reply]
> Lastly, what's the purpose of the `<MenuProvider>` here? Is it required?

It's for submenus (if you're using them). Each submenu controls its own state, which is provided by the `MenuProvider` component. See [Anatomy of nested menus](https://ariakit.org/examples/menu-nested#anatomy-of-nested-menus). If you're using `Menubar` without submenus, you don't need it because `Menubar` already provides the state for its items.

> And on a bit of an off-topic, but might be helpful to others - would [your recommendation](https://github.com/ariakit/ariakit/discussions/3635#discussioncomment-8932285) still hold true if these top-level items would open or be related to a drawer of nested items?
>
> i.e. the About/Contact/Store/Solutions do a navigation, but also have a related element of sub-pages. In the picture - the Store item has sub-sections you can navigate to.

It depends, how does the user open the drawer?

**Reply by bengry** [reply]
> It's for submenus (if you're using them). Each submenu controls its own state, which is provided by the `MenuProvider` component. See [Anatomy of nested menus](https://ariakit.org/examples/menu-nested#anatomy-of-nested-menus). If you're using `Menubar` without submenus, you don't need it because `Menubar` already provides the state for its items.

That makes sense, thanks.

> It depends, how does the user open the drawer?

The drawer can be either pinned or not (using the `Close/Open drawer` button I have pictured).
If it's not pinned - hovering the top-level items (e.g. `Store`) will slide it into view, and pointer-out will slide it out of view.

There's of course also the challenge of what to do for keyboard navigation - I suppose that in that case, while there's tab-focus on one of the items - the drawer will be open (it's contents change based on the hovered item).
But again, I'm open to suggestions on a better behavior.

**Reply by diegohaz** [reply]
> The drawer can be either pinned or not (using the `Close/Open drawer` button I have pictured). If it's not pinned - hovering the top-level items (e.g. `Store`) will slide it into view, and pointer-out will slide it out of view.
> 
> There's of course also the challenge of what to do for keyboard navigation - I suppose that in that case, while there's tab-focus on one of the items - the drawer will be open (it's contents change based on the hovered item). But again, I'm open to suggestions on a better behavior.

In that case, yes, I'd still suggest using a standard HTML navigation. For the drawer, you can use [Hovercard](https://ariakit.org/components/hovercard) triggered by [`onFocusVisible`](https://ariakit.org/reference/hovercard-anchor#onfocusvisible) and a custom position:

```jsx
function TopLevelLink() {
  const [open, setOpen] = useState(false);
  
  return (
    <HovercardProvider open={open} setOpen={setOpen}>
      <HovercardAnchor href="/store" onFocusVisible={() => setOpen(true)}>
        Store
      </HovercardAnchor>
      <Hovercard
        portal
        // disable the popover position
        updatePosition={() => {}}
      >
        ...
```
